File: test/codegen/py/compiled/todo_app.py
# PORTCULLIS INTERNAL
def _plus_(a):
    return lambda b: a + b

def _minus_(a):
    return lambda b: a - b

def _mult_(a):
    return lambda b: a * b

def _div_(a):
    return lambda b: a / b

def _rem_(a):
    return lambda b: a % b

def _gt_(a):
    return lambda b: 1 if a > b else 0

def _gte_(a):
    return lambda b: 1 if a >= b else 0

def _lt_(a):
    return lambda b: 1 if a < b else 0

def _lte_(a):
    return lambda b: 1 if a <= b else 0

def _eq_(a):
    return lambda b: 1 if a == b else 0

def _cons_(a):
    return lambda b: [a] + b

# USER CODE

# signature: ([Atom [Char]] -> ([[Char]] -> [[Char]]))
def update(tup):
  return lambda todos: (_cons_(tup[1])(todos) if _eq_(APPEND)(tup[0]) else (remove(tup[1])(todos) if _eq_(DONE)(tup[0]) else todos))

# signature: ([Char] -> [Atom [Char]])
def append(todo):
  return (APPEND, todo)

# signature: ([Char] -> [Atom [Char]])
def done(done):
  return (DONE, done)

# signature: ([push.a] -> (push.a -> ([push.a] -> [push.a])))
def push(ys):
  return lambda x: lambda xs: _cons_(x)(concat(xs)(ys))

# signature: ([concat.a] -> ([concat.a] -> [concat.a]))
def concat(xs):
  return lambda ys: (ys if _eq_(xs)([]) else push(ys)(xs[0])(xs[1:]))

# signature: ((_filter.x -> Atom) -> (_filter.x -> ([_filter.x] -> [_filter.x])))
def _filter(f):
  return lambda x: lambda xs: concat(([x] if f(x) else []))(filter(f)(xs))

# signature: ((filter.j -> Atom) -> ([filter.j] -> [filter.j]))
def filter(f):
  return lambda xs: (xs if _eq_(xs)([]) else _filter(f)(xs[0])(xs[1:]))

# signature: (neq.a -> (neq.b -> Atom))
def neq(a):
  return lambda b: _eq_(FALSE)(_eq_(a)(b))

# signature: ([Char] -> ([[Char]] -> [[Char]]))
def remove(todo):
  return lambda todos: filter(neq(todo))(todos)

FALSE = 0;
TRUE = 1;
APPEND = 2;
DONE = 3;


pipes = [
  (update, [("&update", 100), ("&todo", 1)], "&todo"),
  (append, [("&append", 50)], "&update"),
  (done, [("&done", 50)], "&update")
]

File: test/codegen/lua/compiled/todo_app.lua
-- PORTCULLIS INTERNAL
function _plus_(a)
  return function (b)
    return a + b
  end
end

function _minus_(a)
  return function (b)
    return a - b
  end
end

function _mult_(a)
  return function (b)
    return a * b
  end
end

function _div_(a)
  return function (b)
    return a / b
  end
end

function _rem_(a)
  return function (b)
    return a % b
  end
end

function _gt_(a)
  return function (b)
    if a > b then
      return 1
    else
      return 0
    end
  end
end

function _gte_(a)
  return function (b)
    if a >= b then
      return 1
    else
      return 0
    end
  end
end

function _lt_(a)
  return function (b)
    if a < b then
      return 1
    else
      return 0
    end
  end
end

function _lte_(a)
  return function (b)
    if a <= b then
      return 1
    else
      return 0
    end
  end
end

function _eq_(a)
  return function (b)
    if a == b then
      return 1
    elseif type(a) == 'table' and type(b) == 'table' then
      if _eq_(a.head)(b.head) == 1 and _eq_(a.tail)(b.tail) == 1 then
        return 1
      else
        return 0
      end
    else
      return 0
    end
  end
end

function _cons_(a)
  return function (b)
    return { head = a, tail = b }
  end
end

function printList(tbl)
  local head = tbl.head
  if head ~= nil then
    if tbl.tail.head ~= nil then
      io.write(head, ', ')
      printList(tbl.tail)
    else
      print(head)
    end
  end
end

-- USER CODE

-- signature: ({Atom [Char]} -> ([[Char]] -> [[Char]]))
function update(tup)
  return function (todos)
    return (_eq_(APPEND)(tup[1]) > 0 and _cons_(tup[2])(todos) or (_eq_(DONE)(tup[1]) > 0 and remove(tup[2])(todos) or todos))
  end
end

-- signature: ([Char] -> {Atom [Char]})
function append(todo)
  return {APPEND, todo}
end

-- signature: ([Char] -> {Atom [Char]})
function done(done)
  return {DONE, done}
end

-- signature: ([push.a] -> (push.a -> ([push.a] -> [push.a])))
function push(ys)
  return function (x)
    return function (xs)
      return _cons_(x)(concat(xs)(ys))
    end
  end
end

-- signature: ([concat.a] -> ([concat.a] -> [concat.a]))
function concat(xs)
  return function (ys)
    return (_eq_(xs)({}) > 0 and ys or push(ys)(xs.head)(xs.tail))
  end
end

-- signature: ((_filter.x -> Atom) -> (_filter.x -> ([_filter.x] -> [_filter.x])))
function _filter(f)
  return function (x)
    return function (xs)
      return concat((f(x) > 0 and { head = x, tail = {} } or {}))(filter(f)(xs))
    end
  end
end

-- signature: ((filter.j -> Atom) -> ([filter.j] -> [filter.j]))
function filter(f)
  return function (xs)
    return (_eq_(xs)({}) > 0 and xs or _filter(f)(xs.head)(xs.tail))
  end
end

-- signature: (neq.a -> (neq.b -> Atom))
function neq(a)
  return function (b)
    return _eq_(FALSE)(_eq_(a)(b))
  end
end

-- signature: ([Char] -> ([[Char]] -> [[Char]]))
function remove(todo)
  return function (todos)
    return filter(neq(todo))(todos)
  end
end

FALSE = 0;
TRUE = 1;
APPEND = 2;
DONE = 3;


pipes = {
  {update, {{"&update", 100}, {"&todo", 1}}, "&todo"},
  {append, {{"&append", 50}}, "&update"},
  {done, {{"&done", 50}}, "&update"}
}
